### PR TITLE
Microchip: MEC172x: Update pinmux driver

### PR DESCRIPTION
**Pull request by scottwcpg**
Update the Microchip XEC pinmux driver to use system I/O
routine for read/write of registers instead of direct use
of volatile and CMSIS defines. Add GPIO port number to
bindings instead of using hard coded value from chip headers.

Signed-off-by: Scott Worley <scott.worley@microchip.com>

### CODE REVIEW

**Comment by scottwcpg** [comment]
Please, please, please do not ask me to use DT FOREACH.  The DT FOREACH macros are not working for this pinmux implementation. I tried for two days and failed.  Don't know why pinmux was not located under soc by the original author(s).

**Comment by scottwcpg** [comment]
Rebased because I missed updating MEC152x DTSI pinmux with the new property.

File: boards/arm/mec172xevb_assy6906/pinmux.c [hunk 1]
```diff
@@ -9,9 +9,105 @@
 #include <kernel.h>
 #include <drivers/pinmux.h>
 
-#include "soc.h"
+#include <soc.h>
```
Comment (jackrosenthal) [inline boards/arm/mec172xevb_assy6906/pinmux.c:12]: please make unrelated style changes in a different PR

File: boards/arm/mec172xevb_assy6906/pinmux.c [hunk 2]
```diff
@@ -9,9 +9,105 @@
...
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_140_176), okay)
+	const struct device *portd;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_200_236), okay)
+	const struct device *porte;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_240_276), okay)
+	const struct device *portf;
+#endif
```
Comment (jackrosenthal) [inline boards/arm/mec172xevb_assy6906/pinmux.c:38]: So I do think you can use the foreach macros. Can you upload the CL where you tried it and we can help debug what you're doing wrong here?

**Review by jackrosenthal — CHANGES_REQUESTED** [review]

File: boards/arm/mec172xevb_assy6906/pinmux.c (hunk 2 above)
Comment (scottwcpg) [reply boards/arm/mec172xevb_assy6906/pinmux.c:38]: @jackrosenthal Latest push contains commented out DT_FOREACH experiments that generate errors.  When I look at other chips for example Atmel and RISC-V I see pinmux under soc with pinmux-someid: pinmux@some-address {};  For Atmel samd2x pinmux and gpio have the same addresses but the DT compiler does not complain.  If I try what Atmel did I get a DTC warning about overlapping addresses.  The HW GPIO pin multiplex control is in the same register as most of the other GPIO configuration.  This means two drivers touching the same registers.  This may explain why the original author of pinmux implemented the way it is.  Most other SoC's have separate HW registers for pin multiplexing and other pin features.

**Comment by scottwcpg** [comment]
@jackrosenthal Should I ask about PINCTRL?  Is the API finalized?

File: drivers/pinmux/pinmux_mchp_xec.c [hunk 3]
```diff
@@ -123,11 +128,82 @@ static const struct pinmux_driver_api pinmux_xec_driver_api = {
...
+ *	};
+ * };
+ * ...
+ * soc {
+ * ...
+ * };
+
+ * How do we use DT_FOREACH_xxxx when pinmux is not under soc, is not in
+ * the form instance_name: class_name@address { properties }; ?
+ */
```
Comment (jackrosenthal) [inline drivers/pinmux/pinmux_mchp_xec.c:157]: You can use DT_INST to get the pinux list, since you have the compatible string "microchip,xec-pinmux", and then DT_FOREAHC_CHILD... at that point to iterate thru the children

Comment (scottwcpg) [reply drivers/pinmux/pinmux_mchp_xec.c]: @jackrosenthal Found the issue.  DT macro failures caused by chip DTSI syntax:
pinmux { ... };
DT macros require by guess work looking at many other SoC's
pinmux: pinmux { ... };

File: dts/arm/microchip/mec1501hsz.dtsi [hunk 4]
```diff
@@ -37,26 +37,32 @@
 		i2c-smb-4 = &i2c_smb_4;
 	};
 
-	pinmux {
+	pinmux: pinmux {
```
Comment (mbolivar-nordic) [inline dts/arm/microchip/mec1501hsz.dtsi:40]: If this node's path in the devicetree (here `/pinmux`) is always going to be the same, you can use `DT_PATH(pinmux)` in C instead of having to define the node label. Then you can remove this change. But adding a node label here is also fine.

Comment (scottwcpg) [reply dts/arm/microchip/mec1501hsz.dtsi:40]: @mbolivar Thanks for the infomation.  I will keep it as is with the node label.  PINCTRL will cause most of this to change.  It appears almost all other SoC have moved to a partial PINCTRL implementation.

**Review by jackrosenthal — APPROVED** [review]

**Review by albertofloyd — APPROVED** [review]
